### PR TITLE
Microsoft.Data.Sqlite: Add incremental blob I/O

### LINKED ISSUE #13824: Microsoft.Data.Sqlite: Incremental blob I/O
See [Open A BLOB For Incremental I/O](https://www.sqlite.org/c3ref/blob_open.html).

This would light up the `GetBytes`, `GetChars` & `GetStream` methods on `SqliteDataReader`.

### PR DESCRIPTION
**Pull request by AlexanderTaeschner**
Adds incremental blob I/O to Microsoft.Data.Sqlite. 
Already presented as PR https://github.com/aspnet/Microsoft.Data.Sqlite/pull/501.
Fixes #13824

### CODE REVIEW

File: src/Microsoft.Data.Sqlite.Core/SqliteBlob.cs [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+using System;
+using System.IO;
+using SQLitePCL;
+
+namespace Microsoft.Data.Sqlite
+{
+    /// <summary>
+    /// Provides methods to access the contents of a BLOB.
+    /// </summary>
+    public class SqliteBlob : Stream
```
Comment (ilmax) [inline src/Microsoft.Data.Sqlite.Core/SqliteBlob.cs]: Should this be called `SqliteBlobStream`?

Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteBlob.cs]: @ilmax We try to match System.Data.SQLite type names where it makes sense to ease the transition.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs [hunk 2]
```diff
@@ -195,8 +197,77 @@ public virtual long GetChars(int ordinal, long dataOffset, char[] buffer, int bu
...
-            => new MemoryStream(GetCachedBlob(ordinal), false);
+        public virtual Stream GetStream(int ordinal, bool writable)
+        {
+            if (ordinal < 0 || ordinal >= FieldCount)
+            {
+                throw new ArgumentOutOfRangeException(nameof(ordinal), ordinal, message: null);
+            }
+
+            var type = raw.sqlite3_column_type(_stmt, ordinal);
+            if ((type != raw.SQLITE_BLOB) || (FieldCount == 1))
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: Why disallow non-BLOBs?

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs [hunk 3]
```diff
@@ -195,8 +197,77 @@ public virtual long GetChars(int ordinal, long dataOffset, char[] buffer, int bu
...
+                if ((dataType == "INTEGER") && (primaryKey != 0))
+                {
+                    rowidOrdinal = i;
+                    break;
+                }
+            }
+
+            if (rowidOrdinal < 0)
+            {
+                return new MemoryStream(GetCachedBlob(ordinal), false);
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: In theory, if we had all the values of the primary key or a unique constraint we *could* SELECT the rowid. But let's not. 😉

File: src/Microsoft.Data.Sqlite.Core/SqliteBlob.cs [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+    {
+        private const string MainDatabaseName = "main";
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteBlob.cs]: Can make the one in SqliteConnection `internal`.

File: test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs [hunk 5]
```diff
@@ -234,6 +234,57 @@ public void Open_works_when_memory_shared()
             }
         }
 
+        [Fact]
+        public void OpenBlob_works()
```
Comment (bricelam) [inline test/Microsoft.Data.Sqlite.Tests/SqliteConnectionTest.cs]: This tests too much--failures are less meaningful. I'll split these out into a new SqliteBlobTests class as I review (and possibly add to) the coverage.

File: src/Microsoft.Data.Sqlite.Core/SqliteBlob.cs [hunk 6]
```diff
@@ -0,0 +1,233 @@
...
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SqliteBlob"/> class.
+        /// </summary>
+        /// <param name="connection">Sqlite connection to be used.</param>
+        /// <param name="tableName">The table name.</param>
+        /// <param name="columnName">The column name.</param>
+        /// <param name="rowid">The row index.</param>
+        /// <param name="writable">Flag indicating whether the blob can be written to.</param>
+        /// <returns>Object to access the BLOB.</returns>
+        public SqliteBlob(SqliteConnection connection, string tableName, string columnName, long rowid, bool writable) :
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteBlob.cs]: It probably makes sense for writable to have a default value of `false` here (and below)

**Comment by bricelam** [comment]
Love it! I just want to bash on SqliteBlob a bit more before merging.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs (hunk 2 above)
Comment (AlexanderTaeschner) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: I removed the condition and added more tests around it (I was not sure if sqlite3_blob_open works on non-BLOBs, but they are handled as TEXT).

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs (hunk 3 above)
Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: This should throw if writable is true

Comment (AlexanderTaeschner) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: It throws now.

**Comment by bricelam** [comment]
Added a sample and some more tests. I want to to a bit more cleanup before merging...

File: test/Microsoft.Data.Sqlite.Tests/SqliteBlobTest.cs [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+        {
+            using (var stream = CreateStream(writable: true))
+            {
+                var ex = Assert.Throws<ArgumentNullException>(
+                    () => stream.Write(null, 0, 0));
+                Assert.Equal("buffer", ex.ParamName);
+            }
+        }
+
+        [Fact] // Not MemoryStream
```
Comment (bricelam) [inline test/Microsoft.Data.Sqlite.Tests/SqliteBlobTest.cs]: Oops, I forgot to remove these comments...

File: src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs [hunk 8]
```diff
@@ -412,7 +412,16 @@ public override long GetChars(int ordinal, long dataOffset, char[] buffer, int b
...
+            => GetStream(ordinal, false);
+
+        /// <summary>
+        ///     Retrieves data as a Stream.
+        /// </summary>
+        /// <param name="ordinal">The zero-based column ordinal.</param>
+        /// <param name="writable">Flag indicating, whether the stream can be written to.</param>
+        /// <returns>The returned object.</returns>
+        public virtual Stream GetStream(int ordinal, bool writable)
+            => _record.GetStream(ordinal, writable);
```
Comment (bricelam) [inline src/Microsoft.Data.Sqlite.Core/SqliteDataReader.cs]: I'm not sure about this overload anymore... I'll discuss with the team.

File: src/Microsoft.Data.Sqlite.Core/SqliteBlob.cs (hunk 6 above)
Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteBlob.cs]: ...or maybe `true`.

File: src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs [hunk 9]
```diff
@@ -195,8 +197,77 @@ public virtual long GetChars(int ordinal, long dataOffset, char[] buffer, int bu
...
+            var type = raw.sqlite3_column_type(_stmt, ordinal);
+            if ((type != raw.SQLITE_BLOB) || (FieldCount == 1))
+            {
+                return new MemoryStream(GetCachedBlob(ordinal), false);
+            }
+
+            var blobDatabaseName = raw.sqlite3_column_database_name(_stmt, ordinal);
+            var blobTableName = raw.sqlite3_column_table_name(_stmt, ordinal);
+
+            var rowidOrdinal = -1;
```
[the remark this answers is not in the document]
Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteDataRecord.cs]: Filed #13987

File: src/Microsoft.Data.Sqlite.Core/SqliteBlob.cs (hunk 6 above)
Comment (bricelam) [reply src/Microsoft.Data.Sqlite.Core/SqliteBlob.cs]: Will discuss when we review the API

**Comment by bricelam** [comment]
(Rebased and squashed on top of master)

**Pull request merged**